core/src/main/kotlin/com/exactpro/th2/read/db/impl/grpc/DataBaseReaderGrpcServer.kt

```diff
@@ -168,31 +172,64 @@ class DataBaseReaderGrpcServer(
     }
 
     private class GrpcExecuteListener(
-        private val observer: StreamObserver<QueryResponse>,
+        streamObserver: StreamObserver<QueryResponse>,
+        private val executionId: Long,
         private val event: Event,
         private val onEvent: (Event) -> Unit,
     ) : ResultListener {
-        override fun onRow(sourceId: DataSourceId, row: TableRow) {
-            requireNotNull(row.executionId) {
-                "'Execution id' is null for row: $row"
+        private val channel = Channel<Any>(1)
+        private val observer = streamObserver as ServerCallStreamObserver<QueryResponse>
+
+        init {
+            observer.setOnReadyHandler {
+                channel.trySend(0)
+            }
+            observer.setOnCancelHandler {
+                LOGGER.warn { "gRPC request is canceled for '$executionId' execution" }
+                channel.close()
+            }
+        }
+
+        override suspend fun onRow(sourceId: DataSourceId, row: TableRow) {
+            check(executionId == row.executionId) {
+                "'Execution id' isn't equal to '$executionId', row: $row"
+            }
+            try {
+                if (observer.isCancelled) {
+                    return
+                }
+                if (!observer.isReady) {
+                    channel.receive()
+                }
+
+                observer.onNext(
+                    QueryResponse.newBuilder()
+                        .putRows(row)
+                        .setExecutionId(row.executionId)
+                        .build()
+                )
+            } catch (e: RuntimeException) {
+                if (e is ClosedReceiveChannelException || e is StatusRuntimeException) {
```

Comment:
thought: Do we need to handle it here? I would think that if we cannot send anything to grpc there is no point in continuing the current request. Unless of course, if we want to save messages to the Cradle for some reason...

Reply:
It is integrating question. Previous implementation is
```
            runCatching { listener.onRow(sourceId, this) }.onFailure {
                LOGGER.error(it) { "error during row processing by listener ${listener::class}" }
            }
```
Let's postpone this question to future releases